### PR TITLE
Add connectEnd attribute to PerformanceResourceTiming interface

### LINKED ISSUE #21262: PerformanceResourceTiming: connectEnd
If a persistent connection is used, connect_end will be set in components/net/http_loader.rs::http_fetch, otherwise, it will be in components/net/http_loader.rs::obtain_response

On getting, the connectEnd attribute MUST return as follows:

1. The same value as fetchStart, if a persistent connection [RFC7230] is used or the resource is retrieved from relevant application caches or local resources.
2. The time immediately after the user agent finish establishing the connection to the server to retrieve the resource, if the last non-redirected fetch of the resource passes the timing allow check algorithm.
  - The returned time MUST include the time interval to establish the transport connection, as well as other time intervals such as SSL handshake and SOCKS authentication.
  - If the transport connection fails and the user agent reopens a connection, connectEnd SHOULD return the corresponding value of the new connection.
3. zero, otherwise.

https://w3c.github.io/resource-timing/#dom-performanceresourcetiming-connectend

### PR DESCRIPTION
**Pull request by pylbrecht**
---

- [x] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [x] These changes fix #21262

- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23145)

### CODE REVIEW

**Comment by pylbrecht** [comment]
Do the wpt tests in `tests/wpt/web-platform-tests/resource-timing/` apply here?

**Comment by pylbrecht** [comment]
Should I update the wpt tests' expectations listed [here](https://build.servo.org/builders/linux-rel-wpt/builds/11934/steps/shell__4/logs/filtered-wpt-errorsummary.log)?

**Comment by paulrouget** [comment]
> Should I update the wpt tests' expectations listed here?

Yes.

**Comment by paulrouget** [comment]
See https://github.com/servo/servo/blob/master/tests/wpt/README.md#updating-test-expectations

File: components/script/dom/performanceresourcetiming.rs [hunk 1]
```diff
@@ -194,4 +194,9 @@ impl PerformanceResourceTimingMethods for PerformanceResourceTiming {
     fn FetchStart(&self) -> DOMHighResTimeStamp {
         Finite::wrap(self.fetch_start)
     }
+
+    // https://w3c.github.io/resource-timing/#dom-performanceresourcetiming-connectend
+    fn ConnectEnd(&self) -> DOMHighResTimeStamp {
+        Finite::wrap(self.connect_end)
```
Comment (jdm) [inline components/script/dom/performanceresourcetiming.rs]: You will need to update PerformanceResourceTiming::from_resource_timing as well to initialize this member.

File: components/net/http_loader.rs [hunk 2]
```diff
@@ -402,6 +403,11 @@ fn obtain_response(
 
     //TODO(#21262) connect_end
```
Comment (jdm) [inline components/net/http_loader.rs]: We can remove this comment now.

**Review by jdm — CHANGES_REQUESTED** [review]
Almost there!

**Comment by pylbrecht** [comment]
@jdm, is this PR in a "finished" state for now? Meaning can I start working on #23150 or does this PR need more attention beforehand?

**Comment by paulrouget** [comment]
@pylbrecht this will be finished once:

1. you get a "r+" (positive review, changes got accepted) from @jdm 
2. all tests pass

As you can see above, one test failed. I could be an intermittent (unrelated test that fails randomly) or a test related to your work. Look at the failing test (/fetch/content-type/script.window.html) and see if it broke because of your code or not (run it locally).

If you think it's unrelated to your work, we will re-run the test for you.

Once the tests pass, and you get a r+, this PR will be merged automatically.

**Comment by paulrouget** [comment]
Also, @jdm mentioned that you should update `PerformanceResourceTiming::from_resource_timing`. Look at that function, you'll see that `connect_end` is initialized to `0` when it should be initialized with your newly introduced `ResourceFetchTiming::connect_end` field.

File: components/script/dom/performanceresourcetiming.rs (hunk 1 above)
Comment (pylbrecht) [reply components/script/dom/performanceresourcetiming.rs]: Fixed in c4bc87709cd8b4f8f510960d996c33b7a29f2c62.

**Comment by pylbrecht** [comment]
Sorry for not elaborating, but @jdm and me have been discussing this issue on IRC yesterday.

@jdm filed an issue for the (intermittent) test failure. So that's not an issue anymore.
Although he discovered another failing test, for which he filed #23150.

> Also, @jdm mentioned that you should update `PerformanceResourceTiming::from_resource_timing`. Look at that function, you'll see that `connect_end` is initialized to `0` when it should be initialized with your newly introduced `ResourceFetchTiming::connect_end` field.

I fixed this in c4bc87709cd8b4f8f510960d996c33b7a29f2c62.

**Comment by jdm** [comment]
@pylbrecht I'm happy to merge this now, but it will require undoing the failing test result from /resource-timing/idlharness.any.worker.html (and squashing the commits).

**Comment by pylbrecht** [comment]
I can't make any sense of the failed AppVeyor build log. Seems unrelated to my changes, but don't hesitate to enlighten me (once again), if I'm wrong.

**Comment by jdm** [comment]
The appveyor failure is a known intermittent crash during compilation.

**Comment by pylbrecht** [comment]
Thanks for the helpful feedback! I hope I got it right this time.

**Comment by CYBAI** [comment]
One more test needs to be labeled as `PASS` 👀

```
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "connectEnd should be 0 in cross-origin request.", 
    "test": "/resource-timing/resource_TAO_zero.htm", 
    "line": 217612, 
    "action": "test_result", 
    "expected": "FAIL"
}
```

**Comment by pylbrecht** [comment]
The deed is done. *fingerscrossed*

**Pull request merged**